testing/endpoint_multihop.go

```diff
@@ -103,13 +231,140 @@ func (ep *EndpointM) SetChannelClosed() error {
 }
 
 // UpdateAllClients updates all client states starting from the first single-hop path to the last.
-// ie. self's client state is propogated to the counterparty chain following the multihop channel path.
+// ie. self's client state is propogated from the counterparty chain following the multihop channel path.
+// This should be called on the chain that's about to receive a Msg with a proof.
 func (ep *EndpointM) UpdateAllClients() error {
-	for _, path := range ep.paths {
-		err := path.EndpointA.UpdateClient()
-		if err != nil {
-			return err
-		}
-	}
-	return nil
+	return ep.Counterparty.mChanPath.UpdateClient()
+}
+
+// GetConnectionHops returns the connection hops for the multihop channel.
+func (ep *EndpointM) GetConnectionHops() []string {
+	return ep.mChanPath.GetConnectionHops()
+}
+
+// CounterpartyChannel returns the counterparty channel used in tx Msgs.
+func (ep *EndpointM) CounterpartyChannel() channeltypes.Counterparty {
+	return channeltypes.NewCounterparty(ep.Counterparty.ChannelConfig.PortID, ep.Counterparty.ChannelID)
+}
+
+// QueryChannelProof queries the multihop channel proof on the endpoint chain.
+func (ep *EndpointM) QueryChannelProof() []byte {
+	channel := ep.GetChannel()
+	channelKey := host.ChannelKey(ep.ChannelConfig.PortID, ep.ChannelID)
+	return ep.QueryMultihopProof(
+		channelKey,
+		ep.Chain.Codec.MustMarshal(&channel),
+		fmt.Sprintf("channel %s", ep.ChannelID),
+	)
+}
+
+// QueryPacketProof queries the multihop packet proof on the endpoint chain.
+func (ep *EndpointM) QueryPacketProof(packet *channeltypes.Packet) []byte {
+	packetKey := host.PacketCommitmentKey(packet.GetSourcePort(), packet.GetSourceChannel(), packet.GetSequence())
+	commitment := channeltypes.CommitPacket(ep.Chain.Codec, packet)
+	return ep.QueryMultihopProof(
+		packetKey, commitment,
+		fmt.Sprintf("packet: %s", packet.String()),
+	)
+}
+
+// QueryPacketAcknowledgementProof queries the multihop packet acknowledgement proof on the endpoint chain.
+func (ep *EndpointM) QueryPacketAcknowledgementProof(packet *channeltypes.Packet) []byte {
+	packetKey := host.PacketAcknowledgementKey(packet.GetDestPort(), packet.GetDestChannel(), packet.GetSequence())
+	commitment := channeltypes.CommitAcknowledgement(ibcmock.MockAcknowledgement.Acknowledgement())
+	return ep.QueryMultihopProof(
+		packetKey, commitment,
+		fmt.Sprintf("packet acknowledgement: %s", packet.String()),
+	)
+}
+
+// QueryMultihopProof queries the proof for a key/value on this endpoint, which is verified on the counterparty chain.
+func (ep *EndpointM) QueryMultihopProof(key, expectedValue []byte, name string) []byte {
```

Comment:
Do we really need the expected Value here? Seems strange to require the expected value when querying?

Reply:
I was looking at `QueryProof()` which does not require a value which prompted this question.

Reply:
yea, makes sense. I dropped the expectedValue and made other relevant changes